### ISSUE TITLE
RFC: search syntax overhaul

### ISSUE THREAD

**Issue by zadam** [zadam/trilium#1041]
Hello, I'm planning a search overhaul for 0.44 - both the actual implementation and the interface (search syntax) will change and **I'd like to get some feedback on the syntax changes**. Changes will affect both classic search and also autocomplete search (e.g. in Jump To dialog).

Best way to explain is with examples:

#### Basic fulltext search

```
rings tolkien
```

This basic search just tries to search everywhere to get the maximum amount of relevant results. This works by decomposing search string into individual words and looks for notes which contain all of the words in any of these places:

*   note title path
    *   e.g. note “Lord of the Rings” in the parent note “Tolkien” form note path “Tolkien > Lord of the Rings" and thus will match search “rings tolkien”
*   note content
*   note's attribute name or value

Note will match even if e.g. “rings” is matched in the note's title and “tolkien” in the attribute

#### Exact match

```
"The Lord of the Rings" Tolkien
```

Similar to basic search above but the quoted string won't be tokenized into words - i.e. we're looking for (case insensitive) sub-string match for “The Lord of the Rings” and “Tokien” in note title, content and attribute name/value.

#### Search by labels

```
towers #book
```

This will search for “towers” the same way as in the fulltext search, but notes will additionally need to have “book” label . Note that similar results could be achieved by searching simply for `towers book`

Note how we use # for denoting label book - the proposition is that labels are now denoted with # and relations with ~ to avoid ambiguity (currently both labels and relations are denoted with @).

**A new feature compared to the existing search is that searching by labels and relations will respect attribute inheritance.**

Labels and other advanced filters should appear after the basic fulltext search. Search strings are actually understood to have two (optional) parts:

1.  series of words which should be searched as fulltext everywhere - in above example it's the `towers tolkien` - these are searched in the note path, title and content, attribute names and values
2.  second part of the search string is formally defined search syntax - in the example above this is `#book`

The two parts of the string are separated by the first occurrence of either `#` or `~` character - in the above example `#` is already part of `#book` filter so no extra character is needed.

#### Search by labels and their values

```
#book #publicationYear = 1954
```

The filter for “book” labels just finds notes with this label without caring about the label value while the “publicationYear” filter matches only notes which have this label associated with exact value 1954.

Other available operators are:

*   `>`
*   `<`
*   `>=`
*   `<=`
*   `*=` ends with given string
*   `=*` starts with given string
*   `*=*` sub-string match

Although all labels are technically strings, when the compared value is detected as number then all comparisons are done as numbers.

#### Search by labels with smart values

```
#dateNote >= TODAY-30
```

Search supports so called “smart” values. Example above will match notes with `dateNote` attribute equaling today's date all the way back to 30 days ago.

Other possible smart values are:

*   NOW +- seconds
*   TODAY +- days
*   WEEK +- days
*   MONTH +- months
*   YEAR +- years

#### Search with relations

```
towers ~author.title *=* Tolkien
```

To refresh the concept, “relation” is a directed named connection from one note to another. So e.g. “Lord of the Rings” can have relation named “author” pointing to the note “J. R. R. Tolkien”.

In the example above, we search with fulltext keyword “towers” and in addition matched notes must have “author” relation and the relation target note title must have sub-string Tolkien.

#### Transitive search with relations

```
~author.relations.son.title = 'Christopher Tolkien'
```

This will search for notes which have “author” relation to a note which has a “son” relation to “Christopher Tolkien” note. This situation can be modeled by this note structure:

*   Books
    *   Lord of the Rings
        *   label: “book”
        *   relation: “author” points to “J. R. R. Tolkien” note
*   People
    *   J. R. R. Tolkien
        *   relation “son” points to "Christopher Tolkien" note
    *   Christoper Tolkien

So in this case the filter needs to go through 3 different levels of notes - it starts with “Lord of the Rings”, jumps through relation “author” to the “J. R. R. Tolkien” from which it follows “son” relation to the “Christoper Tolkien” - here the search algorithm compares the title and the equality with expected “Christoper Tolkien” is satisfied so the starting point - “Lord of the Rings” is matched as a search result.

You might notice that the filtering syntax looks perhaps a bit weird since for denoting the “author” relation we use `~` character while for the “son” relation we prepend it with `relations.` prefix.

The reason for that is that searching by simple relations and labels is expected to be quite common and therefore is “optimized” by assigning special characters of `~` and `#` for relations and labels respectively. Canonical/full name of the above search would be like this:

```
# note.relations.author.relations.son.title = 'Christopher Tolkien'
```

So the `~` character is actually a shortcut for `note.relations.` while the `#` is short for `note.labels.`

But what's with the `#` at the beginning - as discussed above, it is necessary to mark where the fulltext search ends and where formal search syntax begins and for that purpose we use `#` or `~`. This works great when you use shortcut syntax since these characters are already part of the label/relation search, but when using full syntax you must include it as an extra parameter. It's a delimiter of the two parts of the search string so it needs to be included only once.

I realize this is sort of complicated / annoying, but I was not able to figure out a better way how to a) mark the split between two parts of the search string and b) keep this super simple / intuitive for the most common cases (e.g. search string `towers #book` does not actually need any explicit separator)

#### Complex boolean search

```
towers # (#publicationDate >= 1954 AND #publicationDate <= 1960) OR ~author.title *= Tolkien
```

Here we see several things - we have fulltext search for word “towers”, then we mark end of fulltext with # character and define a complex expression with parentheses. Here it's necessary to use `#` character explicitly since the first `(` character already needs to be understood as part of the expression and not as the fulltext.

AND is the default logical operator and can be omitted - e.g. `#book #publicationDate=1954` actually means `#book AND #publicationDate=1954`

#### Ancestors, parents and children

```
# note.parents.title = 'Books'
```

Will find all notes who have parent note with name “Book”.

```
# note.parents.parents.title = 'Books'
```

This again works transitively so this will find notes whose parent of parent is named ‘Book’.

```
# note.ancestors.title = 'Books'
```

This is sort of extension of parents - this will find notes which have an ancestor anywhere in their note path (so parent, grand-parent, grand-grand-parent …) with title ‘Book’. This is a nice way how to reduce scope of the search to a particular sub-tree.

```
# note.children.title = 'sub-note'
```

So this works in the other direction and will find notes which have a child called “sub-note”.

You can ask how this works when there's many children, parents and ancestors - the answer is that with these expressions the note is included as a search result if any single of the e.g. parents satisfies the condition. So e.g. note can have 3 parents, but it's enough if only one of them is called ‘Books’.

#### Search with properties

Note has certain properties which can be also used for searching:

*   dateModified
*   dateCreated
*   utcDateModified
*   utcDateCreated
*   isProtected
*   type (text, code…)
*   title (when you want to search specifically the title)
*   labelCount
*   relationCount
*   attributeCount - labelCount + relationCount
*   parentCount
*   childrenCount
*   isArchived

These are accessed through `note.`, e.g.:

```
# note.type = text
```

Note the `#` and especially the fact that it's a separate token - `#note.type` would mean something different (property of label “note”).

#### Order by and limit

```
#author=Tolkien orderBy #publicationDate desc, note.title limit 10
```

Example above will do the following things (in this sequence):

1.  find notes with label author having value “Tolkien”
2.  order the results by publicationDate in descending order (so newest first)
    1.  in case publication date is equal, use note.title as secondary ordering in ascending order
3.  take only the first 10 results

### Migration

Proposed syntax is incompatible with the existing syntax. Search notes store the search strings and as such will need to be migrated manually to the new syntax.

**Comment by nbreitkopf** [comment]
That looks very powerful. Regarding the split of fulltext part and formal syntax part:
Why not require that the following tokens must be wrapped in quotes, i.e. can only be used in an exact match search, if they are meant to be part of the fulltext search: `#`, `~`, `note.`, `(`, `AND`, `OR`, `orderBy`, `limit`

That way, it is always clear when a formal syntax part starts.

**Comment by zadam** [comment]
@nbreitkopf that's a good thinking.

My concern is that e.g. "(", "and" and "or" are relatively commonly used and as such user might want to search for them without intending to "trigger" formal syntax.

e.g. when user enters `tom and jerry` into the search box, they might be surprised this does not do what they probably intended.

However "and" and "or" might be actually easy cases since they never appear at the beginning of the formal syntax so it's never their job to define to split.

`note.` is relatively easy too since it's quite special prefix which is very unlikely to appear in user's search.

`(` is difficult since it can appear at the beginning of the formal syntax part and it can be also quite commonly used in full text searches. You could search e.g. for "Two Towers (1954)"

`orderBy` and `limit` typically don't appear at the beginning of the format query but theoretically can - e.g. `tolkien orderBy note.title limit 10` seems realistic. But still these words are not so common and would present an issue only when used without filters (which would otherwise define the split).

**Comment by nbreitkopf** [comment]
> My concern is that e.g. "(", "and" and "or" are relatively commonly used and as such user might want to search for them without intending to "trigger" formal syntax.

That's true, but I would argue that these cases almost always should involve an exact search. For instance, `and` is so general, that searching for it is only useful in the context of another phrase.
In your example, the search would also match "Tom Jones and Jerry Lewis". What they actually wanted was `"tom and jerry"`. The same argument can be made for the "Two Towers (1954)" example.

To avoid unexpected behavior for the user, you could try to identify "syntax errors" in the search string and warn the user. For instance `and` or `(` may only be followed by a token that is valid in the formal syntax. I think being restrictive here is better than letting the users get away with a lazy `tom and jerry` search, which in most cases is not what they wanted anyway.

Reserving these keywords also allows you to later extend the search syntax to allow a more complex fulltext search without the full formal syntax part, like this:

`("The Lord of the Rings" OR "Two Towers (1954)") AND Tolkien`

**Comment by lerone-b** [comment]
hey Zaddam,

this all looks very good / productive / promising!
thx for these continuous substantial improvements.

just some minor thougths / feedbacks, if any good:


~author.relations.son.title = 'Christopher Tolkien'
- I see the thinking/logic... but also already sense I would never use something like that.
generally I found, even if myself interested in the power of complex searches, there still is a "cut-off" for me in praxis-terms... mostly meaning what I can´t memorize, I won´t use in the longer run...
– but probably some folks will be happy about this! :-D

# note.ancestors.title = 'Books'
this is cool!
- just looking at all this, I couldn´t tell whether there is a coresponiding leverage (transitive level) for 'children'?

of course just reading things and re-tracing them in one´s mind makes evaluations a little airy; but generally my sense after reading this is that (the explanation of) the concept/syntax/logic of the 'stop token' (#) is kind of critical for guidance/understanding/use.
– this also raises the  question: (why) must the 'stop token' be identical to the 'attribute token'?

... just my thoughts.
of course I expect some misunderstandings on my side, given the nature if this; but wanted to send feedback...

thx again!
oliver

**Comment by zadam** [comment]
> – I see the thinking/logic... but also already sense I would never use something like that.  
> generally I found, even if myself interested in the power of complex searches, there still is a "cut-off" for me in praxis-terms... mostly meaning what I can´t memorize, I won´t use in the longer run...

Yes, it's advanced syntax and you shouldn't need it too often. Exactly this use case has been asked for before though (since the current syntax does not allow it).

These advanced use cases are often meant for saved search notes where you need to do some effort upfront, but use it then continuously ...

> – just looking at all this, I couldn´t tell whether there is a coresponiding leverage (transitive level) for 'children'?

There isn't. I thought about it but decided to not implement the opposite case since I don't see many valid use cases for it. But if it's a requested feature it might be added ...

> – this also raises the question: **(why) must the 'stop token' be identical to the 'attribute token'?**

It doesn't have to be identical, it can be anything in theory. I just reused it since '#' as part of label token also acts as a stop token.

**Comment by meichthys** [comment]
May i suggest adding "DAY" or replacing "TODAY" with "DAY" when using the `datecreated` search functionality? From a consistency perspective, it would be logical to try to use "DAY" - At least i find myself doing that quite often.

**Comment by zadam** [comment]
Functionality described in this issue has been implemented and will be released soon in v0.44 so closing ...
